### PR TITLE
Changes to build against future ROCm ( >4.1)

### PR DESCRIPTION
**Pull request by amathews-amd**
Fixes
* Build error 'explicit specialization after instantiation' (https://groups.google.com/g/cupy/c/-L8HPMTy9n8)
* rocFFT -> hipFFT https://github.com/cupy/cupy/issues/4959

### CODE REVIEW

**Comment by amathews-amd** [comment]
@jithunnair-amd @pruthvistony @jeffdaily @sunway513 - please review

File: cupy_setup_build.py
Comment (leofang) [inline cupy_setup_build.py:119]: This line needs to be reverted. `libraries` here indicates the names of the shared libraries that are linked by the builder, so this change is not necessary because there's no `libhipfft.so`, only `librocfft.so`.

File: install/build.py
Comment (leofang) [inline install/build.py:150]: This change is not needed because `hipfft.h` can be found in `/opt/rocm/include`, so the line below is enough to make it locatable.

File: cupy/cuda/cupy_cub.cu
Comment (leofang) [inline cupy/cuda/cupy_cub.cu:67]: Could you install pytest and run this test locally to see it'd fail?
```
pytest tests/cupy_tests/core_tests/test_ndarray_reduction.py -k TestCubReduction
```

**Review by leofang — CHANGES_REQUESTED** [review]
Thanks for looking into this for us @amathews-amd! A few changes are not necessary.

**Comment by emcastillo** [comment]
Thanks for the PR!

File: cupy_setup_build.py
Comment (amathews-amd) [reply cupy_setup_build.py:119]: See
* https://github.com/ROCmSoftwarePlatform/rocFFT/blob/develop/library/include/hipfft.h#L26
* https://github.com/ROCmSoftwarePlatform/hipFFT#transitioning-from-rocfft 

The change is part of ROCm4.1. 

Is 'HIP_VERSION' available in the build scripts already, so we can key off that ?

File: install/build.py
Comment (amathews-amd) [reply install/build.py:150]: See https://github.com/cupy/cupy/pull/5067#discussion_r610272175 above

File: cupy/cuda/cupy_cub.cu
Comment (amathews-amd) [reply cupy/cuda/cupy_cub.cu:67]: Fails with the change on ROCm4.1 (See below) . Any ideas on how to fix ? 

`
cupy_tests/core_tests/test_ndarray_reduction.py ..F...F...F...F...F...F...F...F................................. 

FAILED cupy_tests/core_tests/test_ndarray_reduction.py::TestCubReduction_param_0_{backend='device', order='C', shape=(10,)}::test_cub_min - AssertionError:
FAILED cupy_tests/core_tests/test_ndarray_reduction.py::TestCubReduction_param_1_{backend='device', order='C', shape=(10, 20)}::test_cub_min - AssertionError:
FAILED cupy_tests/core_tests/test_ndarray_reduction.py::TestCubReduction_param_2_{backend='device', order='C', shape=(10, 20, 30)}::test_cub_min - AssertionError:
FAILED cupy_tests/core_tests/test_ndarray_reduction.py::TestCubReduction_param_3_{backend='device', order='C', shape=(10, 20, 30, 40)}::test_cub_min - AssertionError:
FAILED cupy_tests/core_tests/test_ndarray_reduction.py::TestCubReduction_param_4_{backend='device', order='F', shape=(10,)}::test_cub_min - AssertionError:
FAILED cupy_tests/core_tests/test_ndarray_reduction.py::TestCubReduction_param_5_{backend='device', order='F', shape=(10, 20)}::test_cub_min - AssertionError:
FAILED cupy_tests/core_tests/test_ndarray_reduction.py::TestCubReduction_param_6_{backend='device', order='F', shape=(10, 20, 30)}::test_cub_min - AssertionError:
FAILED cupy_tests/core_tests/test_ndarray_reduction.py::TestCubReduction_param_7_{backend='device', order='F', shape=(10, 20, 30, 40)}::test_cub_min - AssertionError:
_______________________
        numpy.testing.assert_allclose(
            cupy.asnumpy(actual), cupy.asnumpy(desired),
>           rtol=rtol, atol=atol, err_msg=err_msg, verbose=verbose)
E       AssertionError:
E       Not equal to tolerance rtol=1e-05, atol=0
E
E       Mismatched elements: 1 / 1 (100%)
E       Max absolute difference: 9.507621
E       Max relative difference: 1.
E        x: array(0.+0.j, dtype=complex64)
E        y: array(3.834415+8.700122j, dtype=complex64)
`

File: cupy_setup_build.py
Comment (emcastillo) [reply cupy_setup_build.py:119]: did the actual .so file changed names?

Comment (leofang) [reply cupy_setup_build.py:119]: @amathews-amd OK. I checked the new hipfft directory structure, and I think the packaging messed up. 

For all other hip* packages, there is a main header e.g. `/opt/rocm-4.1.0/include/hipblas.h` which symlinks to `/opt/rocm-4.1.0/hipblas/include/hipblas.h`. However, at least for Ubuntu's deb `opt/rocm-4.1.0/include` is empty (no symlink to `opt/rocm-4.1.0/hipfft/include/hipfft.h`). The same applies to shared libraries. This is something ROCm/HIP needs to fix, and let's better not make any exception for hipfft since the old way is still working (just a warning at compile time).

Also, this change would break backward compatibility. We've been able to make things work smoothly since ROCm 3.5+ (one of my development machines is still on that!) and I personally disfavor moving as fast as the upstream. Our expected users aren't moving that fast either. It's arguable that everything on HIP is still experimental, but let's do our best here 🙂

Comment (leofang) [reply cupy_setup_build.py:119]: > Is 'HIP_VERSION' available in the build scripts already, so we can key off that ?

Sorry, forgot to reply on this. Yes, we do set `HIP_VERSION` at the build time, but not here but later, which is why I said it'd break the backward compatibility.

File: install/build.py
Comment (leofang) [reply install/build.py:150]: Please see my reply https://github.com/cupy/cupy/pull/5067/files#r610286064.

File: cupy/cuda/cupy_cub.cu
Comment (leofang) [reply cupy/cuda/cupy_cub.cu:67]: @amathews-amd this is why we needed these lines...😓

This is not a proper fix. It breaks NumPy's [lexical comparison rules for complex numbers](https://numpy.org/doc/stable/reference/generated/numpy.sort.html):
> The sort order for complex numbers is lexicographic. If both the real and imaginary parts are non-nan then the order is determined by the real parts except when they are equal, in which case the order is determined by the imaginary parts.

So you can see that all `max`/`min` tests are failing with complex numbers. (I think in this test we test all dtypes as I figured out a way to make CUB work for everything we do!)

No one is happy with the way NumPy handles the comparison of complex numbers, but the upstream effort to change it is stalled, so unfortunately before it lands in NumPy's main branch we must comply...

The root cause of the build failure is reported in https://github.com/ROCm-Developer-Tools/HIP/issues/2247. Basically the bug is introduced in ROCm 4.1.0 and **its impact is beyond the build error here**. Now hip-clang forbids any possibility of specializing `std::numeric_limits<T>` for non-standard `T`, which IMHO is a *series bug against the C++ standard*. The bug itself should really be fixed asap upstream...

I would suggest you to looking into workarounds for us to 1. build successfully and 2. stay at the same level of the NumPy compliance that we have strived very hard to achieve. I don't think I know enough the internal of rocPRIM so this approach was the only way I found 😅 but it has been working very well since ROCm 3.5.0...

File: cupy_setup_build.py
Comment (amathews-amd) [reply cupy_setup_build.py:119]: @emcastillo, yes

@leofang 
Right now, things should still work the old; Just print a 'deprecated' message. I will reverse the changes in the build scripts for now.

Comment (leofang) [reply cupy_setup_build.py:119]: > did the actual .so file changed names?

@emcastillo Sorry, for some reason GitHub didn't show your reply promptly. 

Since ROCm 4.1.0 there is a new `opt/rocm-4.1.0/lib/libhipfft.so` symlinking to `opt/rocm-4.1.0/hipfft/lib/libhipfft.so`.

Comment (leofang) [reply cupy_setup_build.py:119]: > > did the actual .so file changed names?
> 
> @emcastillo Sorry, for some reason GitHub didn't show your reply promptly.
> 
> Since ROCm 4.1.0 there is a new `opt/rocm-4.1.0/lib/libhipfft.so` symlinking to `opt/rocm-4.1.0/hipfft/lib/libhipfft.so`.

OK so for this particular line it's ok to change for ROCm 4.1.0.

File: cupy/cuda/cupy_cub.cu
Comment (leofang) [reply cupy/cuda/cupy_cub.cu:67]: > 2. stay at the same level of the NumPy compliance that we have strived very hard to achieve.

It is up to the team to decide if we would allow an exception to be made. For example, we could disable the code path for complex numbers on ROCm 4.1.0 if there is really no other way to work around...

Comment (sunway513) [reply cupy/cuda/cupy_cub.cu:67]: >Now hip-clang forbids any possibility of specializing std::numeric_limits<T> for non-standard T, which IMHO is a serious bug violating the C++ standard. The bug itself should really be fixed asap upstream...

cc @yxsamliu, would you help review the reported bug and share you insights if any? Thanks..

File: cupy_setup_build.py
Comment (leofang) [reply cupy_setup_build.py:119]: Had a quick chat offline with @kmaehashi, let's keep this so name change here as the team is interested in seeing the support for ROCm 4.1.0 to land. Backward compatibility starts with 4.1.0.

But, @amathews-amd it'd be great if you could kindly ask your team to fix the directory structure (https://github.com/cupy/cupy/pull/5067#discussion_r610286064) 🙂

File: cupy/cuda/cupy_cub.cu
Comment (leofang) [reply cupy/cuda/cupy_cub.cu:67]: @sunway513, thanks, but @yxsamliu already did, see https://github.com/ROCm-Developer-Tools/HIP/issues/2247#issuecomment-808507106.

Comment (sunway513) [reply cupy/cuda/cupy_cub.cu:67]: @leofang aha, thanks for the heads up and glad it's fully triaged! We'll follow up internally for resolution..

Comment (leofang) [reply cupy/cuda/cupy_cub.cu:67]: Thank you @sunway513 (and @yxsamliu for tracking down the problem)! Please keep us posted if you have a decision 🙂

**Comment by leofang** [comment]
To wrap up quickly with the current discussion about hipFFT: 
1. We can keep using the old header `/opt/rocm/include/hipfft.h` (on ROCm 4.1.0, this is installed by `rocfft`, and hopefully in a future version of ROCm this becomes a working symlink to the actual header installed by `hipfft`)
2. For ROCm 4.1.0 onward, we need to link to `/opt/rocm/lib/libhipfft.so`

File: cupy_setup_build.py
Comment (amathews-amd) [reply cupy_setup_build.py:119]: I discussed the directory structure issue internally with Emmett. 
New issue opened on rocFFT to track this : https://github.com/ROCmSoftwarePlatform/rocFFT/issues/312

Comment (leofang) [reply cupy_setup_build.py:119]: Thanks @amathews-amd!

File: cupy/cuda/cupy_cub.cu
Comment (amathews-amd) [reply cupy/cuda/cupy_cub.cu:67]: Adding comment for note:
Solved in trunk: https://github.com/ROCm-Developer-Tools/HIP/issues/2247#issuecomment-816675887

**Comment by amathews-amd** [comment]
> To wrap up quickly with the current discussion about hipFFT:
> 
> 1. We can keep using the old header `/opt/rocm/include/hipfft.h` (on ROCm 4.1.0, this is installed by `rocfft`, and hopefully in a future version of ROCm this becomes a working symlink to the actual header installed by `hipfft`)
> 2. For ROCm 4.1.0 onward, we need to link to `/opt/rocm/lib/libhipfft.so`

Reverted changes in install/build.py  and cupy/cuda/cupy_cub.cu . Code will not build on ROCm4.1 as-is.

**Comment by emcastillo** [comment]
@amathews-amd how is this going 😀
Can we help you in any way?
Pretty excited to see this working!

**Comment by amathews-amd** [comment]
We are waiting on the numeric_limits fix to land on ROCm.

**Comment by kmaehashi** [comment]
Let me close this as the build now works with ROCm 4.2 (#5209). Thanks again @amathews-amd!

**Pull request closed without merging**
